Add accessibility features to fulfillments

tpaksu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR fixes several accessibility issues in the fulfillment drawer (WOOPLUG-5066):

- **Unique IDs**: Action button components (`cancel-link`, `edit-fulfillment-button`, `fulfill-items-button`, `save-draft-button`, `update-button`, `remove-button`) now use `useInstanceId` to generate unique IDs for `aria-describedby` / `id` pairs, preventing duplicate HTML IDs when multiple fulfillments are visible simultaneously.
- **Removed redundant `aria-label`**: Buttons with visible text no longer have `aria-label` that overrides the visible text (WCAG 2.5.3). The accessible name comes from visible text; supplementary descriptions remain via `aria-describedby`. Exception: the modal Cancel button keeps its `aria-label` since "Cancel" alone is too generic.
- **Moved screen-reader text outside buttons**: `<span className="screen-reader-text">` elements are now siblings of `<Button>` instead of children, preventing them from contributing to the button's computed accessible name.
- **Fixed double announcements in item selector**: Removed `aria-live="polite"` from the count label div; `speak()` calls handle screen reader announcements.
- **Replaced `setTimeout` with double `requestAnimationFrame`**: Focus management in `fulfillment-drawer.tsx` and `fulfillment-editor.tsx` no longer relies on fragile `setTimeout(fn, 100)`.
- **Added focus trapping tests**: Tests for Tab wrapping (last-to-first, first-to-last) and focus restoration on drawer close.
- **Added `speak()` tests**: Tests for screen reader announcements in `item-selector` (select all, deselect all) and `shipment-tracking-number-form` (success, error, API failure, manual provider switch).

Closes # .

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://developer.woocommerce.com/docs/contribution/testing/writing-high-quality-testing-instructions/), include your detailed testing instructions:

1. Open an order edit page that has multiple fulfillments visible at the same time.
2. Inspect the DOM: verify that `aria-describedby` and `id` attributes on action buttons (Cancel, Edit fulfillment, Fulfill items, Save as draft, Update, Remove) contain unique suffixes (e.g. `cancel-link-description-0`, `cancel-link-description-1`).
3. Verify that none of the action buttons have an `aria-label` attribute (except the modal Cancel button which should have `aria-label="Cancel removal and close dialog"`).
4. Open the fulfillment drawer, then press Tab repeatedly: focus should cycle within the drawer (last focusable element wraps to first, Shift+Tab wraps from first to last).
5. Close the drawer: focus should return to the element that was focused before the drawer opened.
6. Enable a screen reader (VoiceOver on macOS). Toggle item checkboxes in the fulfillment item selector: verify each toggle announces the item name and selected/deselected status once (not twice).
7. Use the "Select all" checkbox: verify "N items selected" is announced once.
8. Enter a tracking number and click "Find info": verify success/error messages are announced by the screen reader.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

- All 128 unit tests pass across 20 test suites (`npx jest --config client/jest.config.js --testPathPattern="fulfillments"`)
- ESLint passes on all changed files
- Manual keyboard navigation testing in the fulfillment drawer

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Improve fulfillment drawer accessibility: unique ARIA IDs, fix redundant aria-labels, eliminate double screen reader announcements, and replace setTimeout-based focus management with requestAnimationFrame
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

tpaksu committed:
Add ambiguous provider match notice to tracking number lookup form

tpaksu committed:
Merge branch 'trunk' into add/WOOPLUG-5095-notice-for-low-confidence-provider-match

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Add tests

tpaksu committed:
Add accessibility to fulfillment drawer - part 1

tpaksu committed:
Merge branch 'trunk' into add/WOOPLUG-5066-fulfillments-accessibility-requirements

tpaksu committed:
Fix input focus issues, remove duplicate speaks, replace HTML button with React component, add action button descriptions, fix tests and styling

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Fix linter

tpaksu committed:
Fix another linter error

tpaksu committed:
Merge branch 'trunk' into add/WOOPLUG-5066-fulfillments-accessibility-requirements

tpaksu committed:
Fix error created after merge conflict resolution

tpaksu committed:
Merge branch 'trunk' into add/WOOPLUG-5066-fulfillments-accessibility-requirements

woocommercebot committed:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

tpaksu committed:
Fix fulfillment drawer accessibility issues (WOOPLUG-5066)

- Use useInstanceId for unique ARIA IDs on action buttons
- Remove redundant aria-label that overrides visible button text (WCAG 2.5.3)
- Move screen-reader-text spans outside Button to prevent double naming
- Remove aria-live from item selector count div (speak() handles announcements)
- Replace setTimeout with double requestAnimationFrame for focus management
- Add focus trapping and focus restoration tests for the drawer
- Add speak() announcement tests for item-selector and shipment-tracking-number-form

tpaksu committed:
Merge branch 'trunk' into add/WOOPLUG-5066-fulfillments-accessibility-requirements

tpaksu committed:
Fix lint errors in fulfillment test files

- Fix prettier formatting on toMatch() calls
- Use ownerDocument.activeElement instead of document.activeElement

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/customer-notification-form/index.tsx:
```diff
@@ -21,27 +22,41 @@ export default function CustomerNotificationBox( {
 	type: 'fulfill' | 'update' | 'remove';
 } ) {
 	const { notifyCustomer, setNotifyCustomer } = useFulfillmentContext();
+	const toggleRef = createRef< HTMLInputElement >();
```
I think this should be `useRef` instead of `createRef`. `createRef` creates a new ref object on every render, which means `toggleRef` in the `useEffect` dependency array below will cause it to fire on every render, and the ref might not reliably point to the DOM element. switching to `useRef` would give you a stable reference across renders.

once you make the switch, you could also remove `toggleRef` from the `useEffect` dependency array since it'll be stable.

tpaksu replied:
Fixed. Switched to `useRef` and removed `toggleRef` from the `useEffect` dependency array since it's now stable.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer-header.tsx:
```diff
@@ -36,6 +39,7 @@ export default function FulfillmentsDrawerHeader( {
 							setOpenSection( 'order' );
 							onClose();
 						} }
+						aria-label="Close fulfillment drawer"
```
looks like this `aria-label` isn't wrapped in `__()` for internationalization. non-English screen reader users would always hear this in English regardless of their locale. maybe something like:

```suggestion
						aria-label={ __( 'Close fulfillment drawer', 'woocommerce' ) }
```

tpaksu replied:
Good catch! Wrapped in `__()` and added the `@wordpress/i18n` import.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/style.scss:
```diff
@@ -173,6 +173,28 @@
+			justify-content: center;
+
+			&:focus {
+				outline: none;
```
I believe setting `outline: none; box-shadow: none;` on `:focus` removes all visible focus indicators, which would be a WCAG 2.4.7 (Focus Visible) issue. keyboard users wouldn't be able to tell which element has focus.

might be worth either removing these overrides entirely or using `:focus-visible` instead:
```scss
&:focus-visible {
    outline: 2px solid var(--wp-admin-theme-color, #007cba);
    outline-offset: -2px;
}
```

tpaksu replied:
Fixed. Replaced `:focus { outline: none; box-shadow: none; }` with `:focus-visible` and a proper outline across all three instances in the stylesheet (the header button, the expand button, and the stored fulfillment header button).

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/shipment-form/shipment-tracking-number-form.tsx:
```diff
@@ -278,7 +343,15 @@ export default function ShipmentTrackingNumberForm() {
+				<div
+					id="tracking-number-error"
+					role="alert"
+					aria-live="assertive"
```
seems like `role="alert"` already implies `aria-live="assertive"`, so having both on the same element could cause some screen readers to announce the error twice. you could probably remove `aria-live="assertive"` and keep just `role="alert"`.

tpaksu replied:
Removed the explicit `aria-live="assertive"` since `role="alert"` already implies it.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/action-buttons/fulfill-items-button.tsx:
```diff
@@ -56,14 +61,26 @@ export default function FulfillItemsButton( {
+				isBusy={ isExecuting }
+				disabled={ isExecuting }
+				aria-describedby={ descriptionId }
+				aria-live="polite"
```
not 100% sure about this, but I think placing `aria-live="polite"` directly on a `<button>` means text changes will be announced both as a live region update and as a button name change, which could result in double announcements. since the text change from "Fulfill items" to "Fulfilling…" already conveys status through the accessible name, `aria-live` might not be needed here.

tpaksu replied:
Removed. The status updates are already handled via `speak()` calls, so the `aria-live` on the button was unnecessary.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -27,6 +27,84 @@ const FulfillmentDrawer: React.FC< Props > = ( {
+
+				// Focus the drawer container itself after it's fully rendered
+				// This allows natural scrolling and keyboard navigation within
+				requestAnimationFrame( () => {
```
I wonder if the `requestAnimationFrame` IDs should be stored and cancelled in the `useEffect` cleanup function. if the drawer opens and closes quickly, the pending rAF callbacks could fire on potentially detached DOM nodes, and focus would silently fail to be set. maybe something like:

```tsx
useEffect( () => {
    let rafId1: number;
    let rafId2: number;
    if ( isOpen ) {
        // ...
        rafId1 = requestAnimationFrame( () => {
            rafId2 = requestAnimationFrame( () => {
                if ( drawerRef.current ) {
                    drawerRef.current.focus();
                }
            } );
        } );
    }
    // ...
    return () => {
        cancelAnimationFrame( rafId1 );
        cancelAnimationFrame( rafId2 );
    };
}, [ isOpen ] );
```

same would probably apply to the double rAF in `fulfillment-editor.tsx`.

tpaksu replied:
Added rAF cleanup in both the drawer and the editor `useEffect`s. The cleanup function now stores and cancels both rAF IDs.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -27,6 +27,84 @@ const FulfillmentDrawer: React.FC< Props > = ( {
+			}
+		} else if ( previousFocusRef.current ) {
+			// Restore focus to the previously focused element
+			previousFocusRef.current.focus();
```
if the previously focused element was removed from the DOM while the drawer was open (e.g. a dynamically rendered list re-rendered), `.focus()` would silently fail and the user would end up with no focused element. might be worth guarding with `isConnected`:

```suggestion
		} else if ( previousFocusRef.current?.isConnected ) {
			previousFocusRef.current.focus();
```

tpaksu replied:
Added the `isConnected` guard.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/shipment-form/shipment-tracking-number-form.tsx:
```diff
@@ -278,7 +343,15 @@ export default function ShipmentTrackingNumberForm() {
-			{ error && <ErrorLabel error={ error } /> }
+			{ error && (
+				<div
+					id="tracking-number-error"
```
`id="tracking-number-error"`, `id="finding-status"` (line 229), and `id="provider-ambiguity-notice"` (line 295) are hardcoded. if multiple instances of this component render on the same page, the IDs would collide. the action button components in this same PR use `useInstanceId` for unique IDs, so it could be good to follow the same pattern here for consistency.

tpaksu replied:
Well, this is not really possible, as the tracking number form is rendered only once for the active fulfillment (accordion with others rendering null) but still for better safe than sorry, I replaced all three hardcoded IDs (`tracking-number-error`, `finding-status`, `provider-ambiguity-notice`) with `useInstanceId`, consistent with the action button components.

Ferdev reviewed (changes requested):
Great job on this. the `useInstanceId` + `aria-describedby` pattern across action buttons is well thought out, and the focus trap handles both Tab directions correctly. the `speak()` usage with differentiated politeness levels (polite for success, assertive for errors) follows WCAG best practices.

main things that probably need addressing:

1. `createRef` should be `useRef` in `customer-notification-form/index.tsx` (creates a new ref every render)
2. close button `aria-label` in `fulfillment-drawer-header.tsx` needs `__()` wrapping for i18n
3. focus visibility removed on expand button in `style.scss` (WCAG 2.4.7)
4. redundant `aria-live` + `role="alert"` on error container (potential double announcements)
5. `aria-live="polite"` on the fulfill button (same double announcement risk)
6. `requestAnimationFrame` cleanup in `useEffect` (both drawer and editor)
7. hardcoded IDs in `shipment-tracking-number-form.tsx` (could use `useInstanceId` like the other components)
8. guard `previousFocusRef.current.focus()` with `isConnected` check

tpaksu committed:
Address code review feedback for fulfillment accessibility

- Use useRef instead of createRef in CustomerNotificationBox
- Wrap close button aria-label in __() for i18n
- Replace :focus with :focus-visible for WCAG 2.4.7 compliance
- Remove redundant aria-live on role="alert" error container
- Remove aria-live="polite" from fulfill button to avoid double announcements
- Add requestAnimationFrame cleanup in drawer and editor useEffects
- Replace hardcoded IDs with useInstanceId in ShipmentTrackingNumberForm
- Guard focus restoration with isConnected check

tpaksu committed:
Remove aria-label override on fulfillment header and fix decorative image alt

- Remove aria-label from fulfillment editor expand/collapse header so the
  accessible name comes from the heading text (e.g., "Fulfillment #1"),
  giving screen readers fulfillment-specific context. aria-expanded already
  conveys the expand/collapse state.
- Use empty alt on decorative provider icon that is already aria-hidden.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/action-buttons/test/cancel-link.test.js:
```diff
@@ -21,4 +21,51 @@ describe( 'CancelLink component', () => {
+			expect( button ).not.toHaveAttribute( 'aria-label' );
+		} );
+
+		it( 'should have aria-describedby with unique prefix', () => {
+			render( <CancelLink onClick={ () => {} } /> );
+
+			const button = screen.getByRole( 'button' );
+			expect( button.getAttribute( 'aria-describedby' ) ).toMatch(
+				/^cancel-link-description/
+			);
+		} );
+
+		it( 'should have hidden description for screen readers', () => {
+			render( <CancelLink onClick={ () => {} } /> );
+
+			const description = screen.getByText(
+				'Cancels the current operation without saving changes'
+			);
+			expect( description ).toBeInTheDocument();
+			expect( description.getAttribute( 'id' ) ).toMatch(
+				/^cancel-link-description/
+			);
+			expect( description ).toHaveClass( 'screen-reader-text' );
+		} );
```
Not fixing. The regex prefix check works correctly here: `useInstanceId` generates IDs like `cancel-link-description-0`, and since each test renders a single component, there's no risk of mismatched IDs. Replacing `.toMatch(/^prefix/)` with `.startsWith('prefix')` is functionally identical. The tests already verify the intent.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/action-buttons/test/edit-fulfillment-button.test.js:
```diff
@@ -21,4 +21,44 @@ describe( 'EditFulfillmentButton component', () => {
+			expect( button ).not.toHaveAttribute( 'aria-label' );
+		} );
+
+		it( 'should have aria-describedby with unique prefix', () => {
+			render( <EditFulfillmentButton onClick={ () => {} } /> );
+
+			const button = screen.getByRole( 'button' );
+			expect( button.getAttribute( 'aria-describedby' ) ).toMatch(
+				/^edit-fulfillment-description/
+			);
+		} );
+
+		it( 'should have hidden description for screen readers', () => {
+			render( <EditFulfillmentButton onClick={ () => {} } /> );
+
+			const description = screen.getByText(
+				'Opens the fulfillment editor to modify fulfillment details'
+			);
+			expect( description ).toBeInTheDocument();
+			expect( description.getAttribute( 'id' ) ).toMatch(
+				/^edit-fulfillment-description/
+			);
```
Not fixing. Same reasoning as the cancel-link test comment: the regex prefix assertions already verify the `useInstanceId` wiring correctly, and each test renders a single component instance so there's no risk of ID collision.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/fulfillments/fulfillment-editor.tsx:
```diff
@@ -98,13 +118,20 @@ export default function FulfillmentEditor( {
+					expanded
+						? __( 'Collapse fulfillment details', 'woocommerce' )
+						: __( 'Expand fulfillment details', 'woocommerce' )
+				}
```
Fixed. Removed the `aria-label` so the accessible name now comes from the heading text (e.g., "Fulfillment #1") and the status badge, giving screen readers fulfillment-specific context when multiple fulfillments are present. `aria-expanded` already conveys the expand/collapse state, so the generic label was redundant and harmful.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/shipment-form/shipment-manual-entry-form.tsx:
```diff
@@ -27,7 +27,11 @@ const ShippingProviderListItem = ( {
+						src={ item.icon }
+						alt={ item.label }
+						aria-hidden="true"
+					/>
```
Fixed. Changed to `alt=""` since the image is decorative (the provider name is already rendered as text next to it) and `aria-hidden="true"` already hides it from assistive tech.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/style.scss:
```diff
@@ -173,6 +173,28 @@
+			&:focus {
+				outline: none;
+				box-shadow: none;
+			}
```
Already addressed in commit 2815441. All three `:focus` instances in the stylesheet were replaced with `:focus-visible` using a proper outline.

tpaksu committed:
Fix focus trap escape on Shift+Tab and add fallback aria-label to drawer

- Include drawer panel element in Shift+Tab boundary check so focus
  wraps to the last element instead of escaping the trap
- Add fallback aria-label for the dialog so it has an accessible name
  while the header component is loading (aria-labelledby takes
  precedence once the referenced element exists)

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -27,6 +27,92 @@ const FulfillmentDrawer: React.FC< Props > = ( {
+				if ( focusableElements.length === 0 ) return;
+
+				const firstElement = focusableElements[ 0 ] as HTMLElement;
+				const lastElement = focusableElements[
+					focusableElements.length - 1
+				] as HTMLElement;
+				const activeElement = drawerElement.ownerDocument
+					.activeElement as HTMLElement;
+
+				// Shift+Tab: If focus is on first element, move to last
+				if ( event.shiftKey ) {
+					if ( activeElement === firstElement ) {
+						event.preventDefault();
+						lastElement?.focus();
+					}
+				} else if ( activeElement === lastElement ) {
+					// Tab: If focus is on last element, move to first
+					event.preventDefault();
+					firstElement?.focus();
+				}
```
Fixed. Added activeElement === drawerElement to the Shift+Tab boundary check. Since the drawer panel has tabIndex={-1}, it is excluded from the focusable elements query, so Shift+Tab from the panel was escaping the trap entirely.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -35,14 +121,21 @@ const FulfillmentDrawer: React.FC< Props > = ( {
 						'woocommerce-fulfillment-drawer__panel',
 						isOpen ? 'is-open' : 'is-closed',
 					].join( ' ' ) }
+					role="dialog"
+					aria-modal="true"
+					aria-labelledby="fulfillment-drawer-header"
+					aria-hidden={ ! isOpen }
```
Fixed. Added a fallback aria-label on the dialog element. When the header renders and aria-labelledby resolves, it takes precedence per the ARIA spec. During loading (when the header returns null), the fallback provides the accessible name.

tpaksu commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -27,6 +28,95 @@ const FulfillmentDrawer: React.FC< Props > = ( {
+				if ( ! drawerElement ) return;
+
+				const focusableElements = drawerElement.querySelectorAll(
+					'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
+				);
+
+				if ( focusableElements.length === 0 ) return;
+
+				const firstElement = focusableElements[ 0 ] as HTMLElement;
+				const lastElement = focusableElements[
+					focusableElements.length - 1
+				] as HTMLElement;
+				const activeElement = drawerElement.ownerDocument
+					.activeElement as HTMLElement;
+
+				// Shift+Tab: If focus is on first element or the drawer panel itself, move to last
+				if ( event.shiftKey ) {
+					if (
+						activeElement === firstElement ||
+						activeElement === drawerElement
+					) {
+						event.preventDefault();
+						lastElement?.focus();
+					}
+				} else if ( activeElement === lastElement ) {
+					// Tab: If focus is on last element, move to first
+					event.preventDefault();
+					firstElement?.focus();
+				}
+			}
+		};
+
+		if ( isOpen ) {
+			document.addEventListener( 'keydown', handleKeyDown );
+		}
+
+		return () => {
+			document.removeEventListener( 'keydown', handleKeyDown );
+		};
+	}, [ isOpen, onClose ] );
```
The user should be able to scroll the outside realm to be able to select a different fulfillment from the orders table. That's intended.

tpaksu committed:
Use WordPress admin button focus style (box-shadow) instead of outline

Match the standard WordPress admin focus convention: box-shadow with
--wp-admin-border-width-focus and --wp-admin-theme-color, applied via
:focus:not(:disabled). Also fix the 4th instance in card.scss that was
still suppressing focus indicators.

tpaksu committed:
Revert card.scss focus style change (out of scope)

tpaksu committed:
Merge branch 'trunk' into add/WOOPLUG-5066-fulfillments-accessibility-requirements

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/style.scss:
```diff
@@ -40,9 +40,8 @@
-			&:focus {
-				outline: none;
-				box-shadow: none;
+			&:focus:not(:disabled) {
```
I think the response mentioned switching to `:focus-visible`, but the implementation uses `:focus:not(:disabled)`. the difference is that `:focus` also triggers on mouse clicks, so mouse users would see a focus ring when clicking buttons. `:focus-visible` only shows the indicator for keyboard navigation, which avoids that common UX complaint.

```suggestion
			&:focus-visible {
```

(same applies to lines 183 and 297)

tpaksu replied:
Fixed in a0cddc62da.

Switched all three instances (lines 43, 183, 297) from `:focus:not(:disabled)` to `:focus-visible` so the focus ring only shows on keyboard navigation, not mouse clicks.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -35,14 +125,22 @@ const FulfillmentDrawer: React.FC< Props > = ( {
 					].join( ' ' ) }
+					role="dialog"
+					aria-modal="true"
+					aria-label={ __( 'Fulfillment drawer', 'woocommerce' ) }
```
not 100% sure this works as a fallback. per the WAI-ARIA accessible name computation, when both `aria-label` and `aria-labelledby` are present, `aria-labelledby` always wins, even if the referenced element does not exist in the DOM yet. so `aria-label` would be effectively dead code here.

since the header always renders with the drawer, keeping just `aria-labelledby` alone should be fine:

```suggestion
					aria-labelledby="fulfillment-drawer-header"
```

if a fallback is actually needed, it would need to be conditional (only set one or the other based on whether the header has rendered).

tpaksu replied:
Fixed in a0cddc62da.

Removed the `aria-label` since `aria-labelledby` always takes precedence per the accessible name computation spec. Also removed the now-unused `@wordpress/i18n` import.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/customer-notification-form/index.tsx:
```diff
@@ -21,27 +22,41 @@ export default function CustomerNotificationBox( {
+
+	useEffect( () => {
+		if ( toggleRef.current ) {
+			toggleRef.current.ariaLabel =
```
setting `ariaLabel` and `ariaDescription` via direct DOM mutation bypasses React's reconciliation, so values could be silently lost on re-render. also, `ariaDescription` (mapping to `aria-description`) seems to have limited browser support (not in Safari as of early 2026).

a more robust approach might be using `aria-describedby` pointing to the already-rendered `<p>` element with the content strings, which has universal support. for the label, if `ToggleControl` does not accept `aria-label` as a prop, a visually hidden `<label>` associated via `htmlFor` would probably be more resilient.

not blocking, but worth considering for cross-browser reliability.

tpaksu replied:
Fixed in a0cddc62da.

Replaced `ariaDescription` DOM mutation with `aria-describedby` pointing to the existing `<p>` element (given an `id`). This has universal browser support, unlike `aria-description`. Kept `ariaLabel` via DOM mutation since `ToggleControl` does not expose `aria-label` as a pass-through prop, but `ariaLabel` (mapping to `aria-label`) has full browser support.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/user-interface/fulfillment-drawer/fulfillment-drawer.tsx:
```diff
@@ -27,6 +28,95 @@ const FulfillmentDrawer: React.FC< Props > = ( {
+				if ( ! drawerElement ) return;
+
+				const focusableElements = drawerElement.querySelectorAll(
+					'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
```
looks like the focusable elements selector does not filter out disabled elements:

```
'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
```

this means `firstElement` or `lastElement` could be a disabled button/input that can not receive focus, which would break the trap. the same selector in `fulfillment-editor.tsx` already filters `[disabled]`, so it might be good to keep them consistent:

```suggestion
					'button:not([disabled]), [href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), [tabindex]:not([tabindex="-1"]):not([disabled])'
```

tpaksu replied:
Fixed in a0cddc62da.

Added `:not([disabled])` to the focus trap selector, consistent with the `fulfillment-editor.tsx` selector.

Ferdev commented on plugins/woocommerce/client/admin/client/wp-admin-scripts/fulfillments/components/shipment-form/shipment-tracking-number-form.tsx:
```diff
@@ -122,10 +141,19 @@ export default function ShipmentTrackingNumberForm() {
+				'woocommerce'
+			);
+			speak( successMessage, 'polite' );
 		} catch ( err ) {
```
nit: the catch block discards the actual error object, which could make production debugging difficult. adding a `console.error` would be a low-effort improvement:

```tsx
} catch ( err ) {
	// eslint-disable-next-line no-console
	console.error( 'Tracking number lookup failed:', err );
```

tpaksu replied:
Fixed in a0cddc62da.

Added `console.error( 'Tracking number lookup failed:', err )` with the eslint-disable comment for the `no-console` rule.

Ferdev reviewed (commented):
thanks for addressing the feedback! a few more things below.

tpaksu committed:
Address code review feedback

- Switch :focus:not(:disabled) to :focus-visible for keyboard-only focus rings
- Remove dead aria-label from drawer panel (aria-labelledby takes precedence)
- Replace ariaDescription DOM mutation with aria-describedby for universal support
- Filter disabled elements from focus trap selector to prevent trapping on unfocusable elements
- Add console.error in tracking number lookup catch block for debuggability

tpaksu committed:
Update drawer accessibility test for aria-label removal

Ferdev reviewed (approved):
Awesome work, thanks for all the changes 🙇 

We can 🚢 this 👍
